### PR TITLE
Introduce sequence observers

### PR DESCRIPTION
**Pull request by chanikag**
This is to add sequence observers, which gets invoked at the start and end of a sequence execution. Observers can be configured in conf/sequence-observers.xml

```
<observers>
    <observer name="IntegratorSynapseObserver"
             class="org.wso2.micro.integrator.initializer.MicroIntegratorSequenceController"/>
</observers>

```

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    private static Log log = LogFactory.getLog(SequenceFlowObserversLoader.class);
+
+    public static List<SequenceFlowObserver> loadObservers() {
+        List<SequenceFlowObserver> handlers = new ArrayList<>();
+        OMElement observersConfig =
+                MiscellaneousUtil.loadXMLConfig(SynapseConstants.SEQUENCE_OBSERVERS_FILE);
+        if (observersConfig != null) {
+
+            if (!ROOT_Q.equals(observersConfig.getQName())) {
+                handleException("Invalid handler configuration file");
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java:50]: Invalid handler -> Invalid Sequence observer

File: modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+            Iterator iterator = observersConfig.getChildrenWithName(OBSERVER_Q);
+            while (iterator.hasNext()) {
+                OMElement observerElem = (OMElement) iterator.next();
+
+                String name = null;
+                if (observerElem.getAttribute(NAME_ATT) != null) {
+                    name = observerElem.getAttributeValue(NAME_ATT);
+                } else {
+                    handleException("Name not defined in one or more handlers");
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java:61]: handlers -> observers

File: modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                if (observerElem.getAttribute(CLASS_Q) != null) {
+                    String className = observerElem.getAttributeValue(CLASS_Q);
+                    if (!"".equals(className)) {
+                        SequenceFlowObserver observer = createObserver(className);
+                        if (observer != null) {
+                            handlers.add(observer);
+                            observer.setName(name);
+                        }
+                    } else {
+                        handleException("Class name is null for handle name : " + name);
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java:73]: handler name -> observer name

File: modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                    }
+                } else {
+                    handleException("Class name not defined for handler named : " + name);
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java:76]: handler -> observer

File: modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                } else {
+                    handleException("Name not defined in one or more handlers");
+                }
+
+                if (observerElem.getAttribute(CLASS_Q) != null) {
+                    String className = observerElem.getAttributeValue(CLASS_Q);
+                    if (!"".equals(className)) {
+                        SequenceFlowObserver observer = createObserver(className);
+                        if (observer != null) {
+                            handlers.add(observer);
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java:69]: handlers -> observers

File: modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    private static final QName ROOT_Q = new QName("observers");
+    private static final QName OBSERVER_Q = new QName("observer");
+    private static final QName CLASS_Q = new QName("class");
+    private static final QName NAME_ATT = new QName("name");
+
+    private static Log log = LogFactory.getLog(SequenceFlowObserversLoader.class);
+
+    public static List<SequenceFlowObserver> loadObservers() {
+        List<SequenceFlowObserver> handlers = new ArrayList<>();
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java:44]: handlers -> observers

File: modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        try {
+            obj = Class.forName(classFQName).newInstance();
+        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException e) {
+            handleException("Error creating Handler for class name : " + classFQName, e);
+        }
+
+        if (obj instanceof SequenceFlowObserver) {
+            return (SequenceFlowObserver) obj;
+        } else {
+            handleException("Error creating Handler. The Handler should be of type " +
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java:95]: handler -> observer

File: modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+                    "org.apache.synapse.Handler");
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java:96]: org.apache.synapse.SequenceFlowObserversLoader

File: modules/core/src/main/java/org/apache/synapse/mediators/AbstractListMediator.java [hunk 9]
```diff
@@ -109,6 +117,14 @@ public boolean mediate(MessageContext synCtx, int mediatorPosition) {
                     synCtx.setTracingState(myEffectiveTraceState);
                     if (!mediator.mediate(synCtx)) {
                         returnVal = false;
+                        if (i == mediators.size() - 1) {
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/mediators/AbstractListMediator.java:120]: are we notifying the completeness if the last mediator returns a true?

**Review by isudana — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/synapse/mediators/AbstractListMediator.java (hunk 9 above)
Comment (chanikag) [reply modules/core/src/main/java/org/apache/synapse/mediators/AbstractListMediator.java:120]: This is fixed

File: modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+                String name = null;
+                if (observerElem.getAttribute(NAME_ATT) != null) {
+                    name = observerElem.getAttributeValue(NAME_ATT);
+                } else {
+                    handleException("Name not defined in one or more sequence observer");
+                }
+
+                if (observerElem.getAttribute(CLASS_Q) != null) {
+                    String className = observerElem.getAttributeValue(CLASS_Q);
+                    if (!"".equals(className)) {
```
Comment (SanojPunchihewa) [inline modules/core/src/main/java/org/apache/synapse/config/SequenceFlowObserversLoader.java:66]: Shall we use `isNotBlank` to improve code readability
```suggestion
                    if (StringUtils.isNotBlank(className)) {
```

File: modules/core/src/main/java/org/apache/synapse/mediators/AbstractListMediator.java [hunk 11]
```diff
@@ -80,6 +82,12 @@ public boolean mediate(MessageContext synCtx, int mediatorPosition) {
         // to pass it on; else, do nothing -> i.e. let the parents state flow
         setEffectiveTraceState(synCtx);
         int myEffectiveTraceState = synCtx.getTracingState();
+        if (this instanceof SequenceMediator) {
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/mediators/AbstractListMediator.java:85]: here we need to check whether mediator position is 0

File: modules/core/src/main/java/org/apache/synapse/mediators/AbstractListMediator.java [hunk 12]
```diff
@@ -111,6 +119,14 @@ public boolean mediate(MessageContext synCtx, int mediatorPosition) {
                         returnVal = false;
                         break;
                     }
+                    if (i == mediators.size() - 1) {
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/mediators/AbstractListMediator.java:122]: what if the RuntimeStatisticCollector.isStatisticsEnabled() is true? then observers are not notified isn't it?

**Review by isudana — APPROVED** [review]

**Pull request merged**